### PR TITLE
New Module: mongodb schema

### PR DESCRIPTION
**Pull request by rhysmeister**
##### SUMMARY
New module allowing management of collection validator documents with int tests.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
mongodb_schema

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: tests/integration/targets/mongodb_schema/tasks/mongodb_schema.yml [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+           description: "can only be ACTIVE or DISABLED"
+          },
+          year: {
+            bsonType: "int",
+            minimum: 2021,
+            maximum: 3020,
+            exclusiveMaximum: false,
+            description: "must be an integer from 2021 to 3020"
+          }
+      }
```
Comment (cognifloyd) [inline tests/integration/targets/mongodb_schema/tasks/mongodb_schema.yml:174]:
```suggestion
      properties:
        status:
          bsonType: "string"
          enum: ["ACTIVE", "DISABLED"]
          description: "can only be ACTIVE or DISABLED"
        year:
          bsonType: "int"
          minimum: 2021
          maximum: 3020
          exclusiveMaximum: false
          description: "must be an integer from 2021 to 3020"
```

File: tests/integration/targets/mongodb_schema/tasks/mongodb_schema.yml [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
```
Comment (cognifloyd) [inline tests/integration/targets/mongodb_schema/tasks/mongodb_schema.yml:206]:
```suggestion
      properties:
        status:
          bsonType: "string"
          enum: ["ACTIVE", "DISABLED"]
          description: "can only be ACTIVE or DISABLED"
        year:
          bsonType: "int"
          minimum: 2021
          maximum: 3020
          exclusiveMaximum: false
          description: "must be an integer from 2021 to 3020"
```

File: plugins/modules/mongodb_schema.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+      - List of fields that are required.
+    type: list
+    elements: str
+  properties:
+    description:
+      - Individual property specification.
+    type: dict
+  action:
+    description:
+      - The validation action to perform.
```
Comment (cognifloyd) [inline plugins/modules/mongodb_schema.py:55]:
```suggestion
      - The validation action for MongoDB to perform when handling invalid documents.
```

File: plugins/modules/mongodb_schema.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+    type: str
+    choices:
+      - "error"
+      - "warn"
+    default: "error"
+  level:
+    description:
+      - The validation level.
```
Comment (cognifloyd) [inline plugins/modules/mongodb_schema.py:63]:
```suggestion
      - The validation level MongoDB should apply when updating existing documents.
```

File: plugins/modules/mongodb_schema.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+version_added: "1.0.0"
+
+extends_documentation_fragment:
+  - community.mongodb.login_options
+  - community.mongodb.ssl_options
+
+options:
+  db:
+    description:
+      - The database to work with
```
Comment (cognifloyd) [inline plugins/modules/mongodb_schema.py:31]:
```suggestion
      - The database to work with.
```

**Review by cognifloyd — COMMENTED** [review]
Cool stuff! It's exciting to see all the new modules you're adding to work with MongoDB from Ansible.

A few yaml style & doc nitpicks.

Also, do you plan to add all of the `$jsonSchema` options? https://docs.mongodb.com/manual/reference/operator/query/jsonSchema/#available-keywords

**Comment by rhysmeister** [comment]
@cognifloyd Thanks for the feedback. re: Options, I think these woulds pretty much be usable within the properties section without any further work needed. To be honest I don't really know a lot about schema validation in MongoDB so we'll have to see how/if it get used.

File: plugins/modules/mongodb_schema.py [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+  collection:
+    description:
+      - The collection to work with.
+    required: yes
+    type: str
+  bsonType:
+    description:
+      - Expected document type.
+    type: str
+    default: "object"
```
Comment (cognifloyd) [inline plugins/modules/mongodb_schema.py:45]: Can this every be something other than `object`? As I understand MongoDB documents are always objects. Maybe we should just hard code this to be `object`.

**Review by cognifloyd — COMMENTED** [review]
Since documents are always `objects` (right?), these are the possible top-level [`$jsonSchema`](https://docs.mongodb.com/manual/reference/operator/query/jsonSchema/#json-schema) fields on validation (excluding `bsonType`, `required`, and `properties` which you've already added):
- `enum` (array of values)
- `allOf` (array of JSON Schema objects)
- `anyOf` (array of JSON Schema objects)
- `oneOf` (array of JSON Schema objects)
- `not` (a JSON Schema object)
- `maxProperties` (integer)
- `minProperties` (integer)
- `additionalProperties` (boolean or a JSON Schema object)
- `patternProperties` (object like `properties`: object with regex keys and JSON Schema objects for values)
- `dependencies` (object with string keys and for values either: JSON Schema objects or arrays with at least one string)
- `title` (string)
- `description` (string)

File: plugins/modules/mongodb_schema.py (hunk 6 above)
Comment (rhysmeister) [reply plugins/modules/mongodb_schema.py:45]: @cognifloyd Can't find anything confirming this but you're almost certainly right.

**Pull request merged**